### ISSUE TITLE
Feature Request: Ability to set UID/GID after sync

### ISSUE THREAD

**Issue by dsuievalov** [devspace-cloud/devspace#849]
**Is your feature request related to a problem?**  
I like the `devspace` tool since it is fast and is not buggy but it would be nice to have the ability to change UID/GID attributes of files after `sync` to make sure that application that runs in kubernetes from unprivileged user can access them.

**Which solution do you suggest?**  

Add options `dev.sync[*].upstreamUser.uid`, `dev.sync[*].upstreamUser.gid` which will instruct `sync` to change the UID/GID for uploaded files. In addition, please add arguments to a cli version of `devspace sync` to support this feature too (possibly `--upstream-user-uid`,`--upstream-user-gid`).

**Which alternative solutions exist?**  
Change uid/gid manually `kubectl exec -it <pod> -- bash -c 'chown -R <UID>:<GID> <synced_folder>'` which break the idea of auto sync.

**Additional context**  

/kind feature

**Comment by FabianKramm** [comment]
@dsuievalov thanks for opening this issue! I'm actually not sure if we should support this, since this would promote a possible wrong idea how to use the sync with unprivileged user images.  

If you have a container that runs by default with a non root user, the sync will be started with this exact same user and group and hence all upstreamed files and folders will have the same uid and gid and can only be synced from and to the locations where the user has access to. Adding an option such as "upstreamUser.uid" would mean the sync needs to be started as root user in an otherwise non root container (bad!) and only then it would be able to change the uid and gid of the upstreamed files to the desired user.

**Comment by dsuievalov** [comment]
@FabianKramm thanks for the update. Agreed with the fact that it can lead to some misunderstanding if it will not be highlighted good enough in the doc.

My use case is the following: PHP application runs under `myuser` that have enough permissions and the child processes runs as non-interactive/non-privileged `dummy` user. To use `devspace sync/dev` I need to make sure that files on the upstream will get the correct permission for the `dummy` user can access them in RW mode. Since the container is running from `myuser` only it has all the permissions for files that were uploaded. If `dummy` user has the same level of access that `myuser` it will have access to the main php configuration file which is a security issue for me. Which is why it will be nice to have the ability the permissions per specific user:group.

**Comment by FabianKramm** [comment]
@dsuievalov thanks for explaining your use case! I think what we could do, which would also help you in your use case, is to add a config option `dev.sync.command.onUpload` and `dev.sync.command.onDownload`. I think this would have more benefits and use cases than just making an option for changing the UID/GID for files. Would that be okay for you?

**Comment by dsuievalov** [comment]
@FabianKramm Totally agree with you, this feature will be more valuable and cover my case too :)

**Comment by LukasGentele** [comment]
I like the idea of `onUpload/onDownload` but I would either rename `command` to `trigger` or leave out that additionally nesting layer completely. 

For both config options, I would recommend an object structure with the following fields:
- command: string 
- execInContainer: true|false (default: true?)

If execInContainer == false, we should use the same command string interpreter as with custom commands. That would allow for maximum portability.

What do you think? @dsuievalov @FabianKramm

**Comment by LukasGentele** [comment]
Another thing to consider would be: what if I wanted to run one command locally and another one inside container. Not sure if that might be a use case but it would probably be easy to find a config structure to make sure this case is also covered.

**Comment by dsuievalov** [comment]
@LukasGentele I feel that `execInContainer ` is redundant in `onUpload ` case because files will be uploaded into the running container and run in a container in any case. Is that the expected behavior or I miss something?

**Comment by LukasGentele** [comment]
@dsuievalov If you want to run a command locally on your computer after uploading a file, that is not possible without this option. Might be a use case. I know that other sync solutions allow to run commands locally after each sync. 

Not sure if we should limit us to only one runtime (container) for uploads and one runtime (local) for downloads because it is technically easy to implement a more flexible config. However, I strongly agree with you that onUpload already suggests that the command will run in the container while onDownload suggest to run it locally which makes it hard to define a consistent default value for a config option like execInContainer. Any suggestions/alternatives?

**Comment by dsuievalov** [comment]
Have a slack conversation with @LukasGentele and we ended with the following format:

```yaml
dev:
  sync:
    - onUpload:
        exec:
          # if we uploaded files execute locally
          local: bash
          # if we uploaded files execute remotely
          remote: sh
      onDownload:
        exec:
          # if we downloaded files execute locally
          local: bash
          # if we downloaded files execute remotely
          remote: bash
```
@FabianKramm what do you think about the following structure?

**Comment by FabianKramm** [comment]
@dsuievalov is there an important use case for the `onDownload.exec.remote` and `onUpload.exec.local`? The problem with these two is that they would definitely slow down the sync (for each `onDownload.exec.remote` basically a `kubectl exec` has to be run and the sync has to wait for it otherwise the chances are very high that the sync loses track of the file, depending on what the command is doing to it) or could easily lead to endless loops of redownloading / reuploading files depending on what command is triggered (if the sync is not waiting for the command to finish).  

Another question regarding this design is, how does the specified command know which file it should modify? I think I would prefer a separate command & args field, where we could specify an index or placeholder for the file argument.

**Comment by dsuievalov** [comment]
> @dsuievalov is there an important use case for the `onDownload.exec.remote` and `onUpload.exec.local`? The problem with these two is that they would definitely slow down the sync (for each `onDownload.exec.remote` basically a `kubectl exec` has to be run and the sync has to wait for it otherwise the chances are very high that the sync loses track of the file, depending on what the command is doing to it) or could lead to endless loops of redownloading / reuploading files depending on what command is triggered (if the sync is not waiting for the command to finish).

I am more interested in `onDownload.exec.local`/`onUpload.exec.remote` and think most use cases (around 90%) will be covered with these options. 

> Another question regarding this design is, how does the specified command know which file it should modify? I think I would prefer a separate command & args field, where we could specify an arg index or arg placeholder for the file argument instead of doing a string replace.

If I understand the concept correctly `exec` will be triggered when files downloaded/uploaded. But it will not track the files itself - only the action. For my use case, I will just use the find command that will search for files/dirs that are not owned by specific `UID`/`GID`. From the performance side, it is not the best choice but nevertheless. On the other hand, if `devspace` will have the ability to send the list of updated files to the exec it will be easier.

**Comment by dsuievalov** [comment]
@FabianKramm I am wondering if `devspace` can track if it creates a new `dirs` in the same way as a new `files`? If yes, the following way of running `chown` is fine for me. If not then I will need to do something like this:

```
sync:
  - onUpload:
      execRemote:
        command: ["/bin/bash"] # will be run for every changed file
        args: ["-c", "chown newowner:newgroup ${1} && find /synced-dir -type d -not -user newowner -exec chown newowner:newgroup {}\;", "--", ""] # empty argument could be replaced with the actual changed filename, if omitted no filename is used or appended
```
And `find` will be called on each file change.

**Comment by FabianKramm** [comment]
@dsuievalov yes definitely, we would run it for every folder/file created, however I'm not sure if we should allow separate commands for this or just run the same command for both.

**Comment by dsuievalov** [comment]
@FabianKramm I think if `devspace` tracks `dirs` separately from `files` it is better to run the command for each action in separate. In that way, the exec will be triggered extra times. For example:

- we copied from local system the following dir:

```bash
[~]$ ls test/
1    11   14   17   2    22   25   28   30   33   36   39   41   44   47   5    52   55   58   60   63   66   69   71   74   77   8    82   85   88   90   93   96   99
10   12   15   18   20   23   26   29   31   34   37   4    42   45   48   50   53   56   59   61   64   67   7    72   75   78   80   83   86   89   91   94   97
100  13   16   19   21   24   27   3    32   35   38   40   43   46   49   51   54   57   6    62   65   68   70   73   76   79   81   84   87   9    92   95   98
```

- if we run the command separately for `dir` and `files` we will run the command for dir `test` once and ones for each file in that dir;
- if we are not running run the command for `dir` and `files`  separately we will run the command 100 times for dir `test` and ones for each file in that dir;

P.S. Can we specify it cleaner in yaml that we run the command for each change in fs? Something like this:

```yaml
sync:
  - onUpload:
      onEachChange:
        execRemote:
          command: ["chown"]
          args: ["newowner:newgroup", ""]
```

**Comment by dsuievalov** [comment]
Agreed with @FabianKramm on the following syntax:

```yaml
sync:
  - onUpload:
    execRemote:
      # will run for every change with files (`create`/`modify`) and dirs (`create`)
      command: ["chown"] 
      args: ["newowner:newgroup", ""]
      # in addition to general command will run for every change with files (`create`/`modify`)
      onFileChange:
        command: ["echo"]
        args: ["Additional command for the changed files", ""]
      # in addition to general command will run for every change with dirs (`create`)
      onDirCreate:
        command: ["echo"]
        args: ["Additional command for the changed dirs", ""]
# the same for onDownload
```

@LukasGentele are you fine with that?

**Comment by LukasGentele** [comment]
That makes sense to me. Focussing on the more important use cases is great. We should create separate issues for the other two that we'll leave out for now (onUpload.execLocal and onDownload.execRemote). 

The only thing that needs improvement is the "empty arg" = file/dir path logic. To me, that is really unusual and looks like a mistake. It looks even stranger when not using the [] array syntax in yaml but rather separate lines. And I strongly believe that it is unintuitive because no one would understand what is going on here just from seeing the code snippet. I would suggest using something more like xargs oder basb vars, e.g. `$PATH` or `$SYNC_PATH` or `{}` or even appending the path if the var is not used (I think that is what xargs does as well).

**Comment by dsuievalov** [comment]
+1 for `{}` syntax

**Comment by FabianKramm** [comment]
alright then we do `{}` as placeholder for the changed file / folder name

**Comment by FabianKramm** [comment]
Implemented with #868
